overlord/snapstate/aliases.go

```diff
@@ -177,29 +222,49 @@ func (m *SnapManager) doAlias(t *state.Task, _ *tomb.Tomb) error {
 			}
 			return fmt.Errorf("cannot %s alias %q for %q, no such alias", action, alias, snapName)
 		}
-		if aliasStatuses[alias] == newStatus {
-			// nothing to do
-			continue
-		}
-		beAlias := &backend.Alias{
-			Name:   alias,
-			Target: filepath.Base(aliasApp.WrapperPath()),
-		}
+		op := "nop"
 		switch newStatus {
 		case "enabled":
 			if aliasStatuses[alias] != "auto" {
+				op = "add"
+			}
+		case "disabled":
+			if aliasStatuses[alias] != "" {
+				op = "rm"
+			}
+		case "auto":
+			if autoSet[alias] && aliasApp != nil {
+				if aliasStatuses[alias] != "enabled" {
+					op = "add"
+				}
+			} else {
+				if aliasStatuses[alias] == "enabled" {
+					op = "rm"
+				}
+				newStatus = ""
+			}
+		}
+		if newStatus != "" {
+			aliasStatuses[alias] = newStatus
+		} else {
+			delete(aliasStatuses, alias)
+		}
+		if aliasApp != nil && op != "nop" {
```

Comment:
This case, specifically, is a bit unclear. Why do we get this far with a nil `aliasApp`? I'd expect it to be bail almost immediately at the top of the loop, perhaps with a small bit of logic to handle it (adjusting the map). Perhaps this would also avoid the curious nop altogether.